File: backend/algorithms/graphs.py
# Python Graph Algorithms Step Generator
from collections import deque
import heapq

FIXED_NODE_POSITIONS = {
    'A': {'x': 100, 'y': 150},
    'B': {'x': 250, 'y': 80},
    'C': {'x': 250, 'y': 220},
    'D': {'x': 400, 'y': 80},
    'E': {'x': 400, 'y': 220},
    'F': {'x': 550, 'y': 150},
    'G': {'x': 320, 'y': 50},
    'H': {'x': 320, 'y': 250}
}

DEFAULT_GRAPH = {
    'nodes': [
        {'id': 'A', 'x': 100, 'y': 150},
        {'id': 'B', 'x': 250, 'y': 80},
        {'id': 'C', 'x': 250, 'y': 220},
        {'id': 'D', 'x': 400, 'y': 80},
        {'id': 'E', 'x': 400, 'y': 220},
        {'id': 'F', 'x': 550, 'y': 150}
    ],
    'edges': [
        {'from': 'A', 'to': 'B', 'weight': 4},
        {'from': 'A', 'to': 'C', 'weight': 2},
        {'from': 'B', 'to': 'C', 'weight': 1},
        {'from': 'B', 'to': 'D', 'weight': 5},
        {'from': 'C', 'to': 'E', 'weight': 8},
        {'from': 'C', 'to': 'D', 'weight': 10},
        {'from': 'D', 'to': 'E', 'weight': 2},
        {'from': 'D', 'to': 'F', 'weight': 6},
        {'from': 'E', 'to': 'F', 'weight': 3}
    ]
}


def normalize_graph_nodes(graph):
    if not graph or 'nodes' not in graph:
        return DEFAULT_GRAPH['nodes'], DEFAULT_GRAPH['edges']

    nodes = graph['nodes']
    edges = graph['edges']

    # Enforce fixed node positions so graph layout shape NEVER distorts
    for n in nodes:
        node_id = n.get('id')
        if node_id in FIXED_NODE_POSITIONS:
            n['x'] = FIXED_NODE_POSITIONS[node_id]['x']
            n['y'] = FIXED_NODE_POSITIONS[node_id]['y']

    return nodes, edges


def generate_bfs_steps(graph=None, start='A'):
    nodes, edges = normalize_graph_nodes(graph)
    steps = []

    adj = {n['id']: [] for n in nodes}
    for edge in edges:
        if edge['from'] in adj:
            adj[edge['from']].append((edge['to'], edge))
        if edge['to'] in adj:
            adj[edge['to']].append((edge['from'], edge))

    visited = set([start])
    queue = deque([start])
    visited_edges = []

    steps.append({
        'nodes': nodes,
        'edges': edges,
        'currentNode': start,
        'visitedNodes': list(visited),
        'activeEdge': None,
        'visitedEdges': [],
        'queue': list(queue),
        'line': 1,
        'description': f'Start BFS from node {start}'
    })

    while queue:
        curr = queue.popleft()
        steps.append({
            'nodes': nodes,
            'edges': edges,
            'currentNode': curr,
            'visitedNodes': list(visited),
            'activeEdge': None,
            'visitedEdges': list(visited_edges),
            'queue': list(queue),
            'line': 3,
            'description': f'Dequeued node {curr}. Exploring neighbors'
        })

        for neighbor, edge in adj.get(curr, []):
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)
                visited_edges.append(edge)
                steps.append({
                    'nodes': nodes,
                    'edges': edges,
                    'currentNode': neighbor,
                    'visitedNodes': list(visited),
                    'activeEdge': edge,
                    'visitedEdges': list(visited_edges),
                    'queue': list(queue),
                    'line': 6,
                    'description': f'Discovered unvisited neighbor {neighbor}. Enqueued {neighbor}'
                })

    steps.append({
        'nodes': nodes,
        'edges': edges,
        'currentNode': None,
        'visitedNodes': list(visited),
        'activeEdge': None,
        'visitedEdges': list(visited_edges),
        'queue': [],
        'line': 8,
        'description': 'BFS Traversal Complete!'
    })
    return steps


def generate_dfs_steps(graph=None, start='A'):
    nodes, edges = normalize_graph_nodes(graph)
    steps = []

    adj = {n['id']: [] for n in nodes}
    for edge in edges:
        if edge['from'] in adj:
            adj[edge['from']].append((edge['to'], edge))
        if edge['to'] in adj:
            adj[edge['to']].append((edge['from'], edge))

    visited = set()
    visited_edges = []

    steps.append({
        'nodes': nodes,
        'edges': edges,
        'currentNode': start,
        'visitedNodes': [],
        'activeEdge': None,
        'visitedEdges': [],
        'line': 1,
        'description': f'Start DFS from node {start}'
    })

    def dfs(node):
        visited.add(node)
        steps.append({
            'nodes': nodes,
            'edges': edges,
            'currentNode': node,
            'visitedNodes': list(visited),
            'activeEdge': None,
            'visitedEdges': list(visited_edges),
            'line': 2,
            'description': f'Visit node {node}'
        })

        for neighbor, edge in adj.get(node, []):
            if neighbor not in visited:
                visited_edges.append(edge)
                steps.append({
                    'nodes': nodes,
                    'edges': edges,
                    'currentNode': node,
                    'visitedNodes': list(visited),
                    'activeEdge': edge,
                    'visitedEdges': list(visited_edges),
                    'line': 4,
                    'description': f'Traverse edge ({edge["from"]}-{edge["to"]}) to unvisited node {neighbor}'
                })
                dfs(neighbor)

    dfs(start)
    steps.append({
        'nodes': nodes,
        'edges': edges,
        'currentNode': None,
        'visitedNodes': list(visited),
        'activeEdge': None,
        'visitedEdges': list(visited_edges),
        'line': 6,
        'description': 'DFS Traversal Complete!'
    })
    return steps


def generate_dijkstra_steps(graph=None, start='A'):
    nodes, edges = normalize_graph_nodes(graph)
    steps = []

    adj = {n['id']: [] for n in nodes}
    for edge in edges:
        if edge['from'] in adj:
            adj[edge['from']].append((edge['to'], edge['weight'], edge))
        if edge['to'] in adj:
            adj[edge['to']].append((edge['from'], edge['weight'], edge))

    dist = {n['id']: float('inf') for n in nodes}
    dist[start] = 0
    pq = [(0, start)]
    visited = set()
    visited_edges = []

    steps.append({
        'nodes': nodes,
        'edges': edges,
        'currentNode': start,
        'dist': {k: (v if v != float('inf') else '∞') for k, v in dist.items()},
        'visitedNodes': [],
        'activeEdge': None,
        'visitedEdges': [],
        'line': 1,
        'description': f"Start Dijkstra's algorithm from source node {start}"
    })

    while pq:
        d, curr = heapq.heappop(pq)
        if curr in visited:
            continue
        visited.add(curr)

        steps.append({
            'nodes': nodes,
            'edges': edges,
            'currentNode': curr,
            'dist': {k: (v if v != float('inf') else '∞') for k, v in dist.items()},
            'visitedNodes': list(visited),
            'activeEdge': None,
            'visitedEdges': list(visited_edges),
            'line': 3,
            'description': f'Extracted minimum distance node {curr} with distance {d}'
        })

        for neighbor, weight, edge in adj.get(curr, []):
            if neighbor not in visited:
                alt = dist[curr] + weight
                if alt < dist[neighbor]:
                    dist[neighbor] = alt
                    heapq.heappush(pq, (alt, neighbor))
                    visited_edges.append(edge)
                    steps.append({
                        'nodes': nodes,
                        'edges': edges,
                        'currentNode': neighbor,
                        'dist': {k: (v if v != float('inf') else '∞') for k, v in dist.items()},
                        'visitedNodes': list(visited),
                        'activeEdge': edge,
                        'visitedEdges': list(visited_edges),
                        'line': 6,
                        'description': f'Updated shortest distance to {neighbor} via {curr}: {dist[neighbor]}'
                    })

    steps.append({
        'nodes': nodes,
        'edges': edges,
        'currentNode': None,
        'dist': {k: (v if v != float('inf') else '∞') for k, v in dist.items()},
        'visitedNodes': list(visited),
        'activeEdge': None,
        'visitedEdges': list(visited_edges),
        'line': 8,
        'description': "Dijkstra's Shortest Paths Complete!"
    })
    return steps

File: src/algorithms/graphs.js
// Step Generators for Graph Algorithms (BFS, DFS, Dijkstra, A*, Prim's, Kruskal's)

export const DEFAULT_GRAPH = {
  nodes: [
    { id: 'A', x: 100, y: 150 },
    { id: 'B', x: 250, y: 80 },
    { id: 'C', x: 250, y: 220 },
    { id: 'D', x: 400, y: 80 },
    { id: 'E', x: 400, y: 220 },
    { id: 'F', x: 550, y: 150 }
  ],
  edges: [
    { from: 'A', to: 'B', weight: 4 },
    { from: 'A', to: 'C', weight: 2 },
    { from: 'B', to: 'C', weight: 1 },
    { from: 'B', to: 'D', weight: 5 },
    { from: 'C', to: 'E', weight: 8 },
    { from: 'C', to: 'D', weight: 10 },
    { from: 'D', to: 'E', weight: 2 },
    { from: 'D', to: 'F', weight: 6 },
    { from: 'E', to: 'F', weight: 3 }
  ]
}

export function generateRandomGraph() {
  const labels = ['A', 'B', 'C', 'D', 'E', 'F']
  const nodes = labels.map((id, idx) => {
    const angle = (idx / labels.length) * 2 * Math.PI
    return {
      id,
      x: 320 + 160 * Math.cos(angle),
      y: 180 + 110 * Math.sin(angle)
    }
  })

  const edges = []
  for (let i = 0; i < labels.length; i++) {
    const next = (i + 1) % labels.length
    edges.push({
      from: labels[i],
      to: labels[next],
      weight: Math.floor(Math.random() * 9) + 1
    })
  }

  // Cross edges
  edges.push({ from: 'A', to: 'C', weight: Math.floor(Math.random() * 9) + 1 })
  edges.push({ from: 'B', to: 'E', weight: Math.floor(Math.random() * 9) + 1 })
  edges.push({ from: 'D', to: 'F', weight: Math.floor(Math.random() * 9) + 1 })

  return { nodes, edges }
}

export function generateBFSGraphSteps(graph = DEFAULT_GRAPH, startNode = 'A') {
  const steps = []
  const visited = new Set([startNode])
  const queue = [startNode]
  const visitedEdges = []

  steps.push({
    nodes: graph.nodes,
    edges: graph.edges,
    currentNode: startNode,
    visitedNodes: Array.from(visited),
    visitedEdges: [...visitedEdges],
    queue: [...queue],
    line: 1,
    description: `Start BFS graph traversal from source node ${startNode}`
  })

  const adj = {}
  graph.nodes.forEach(n => { adj[n.id] = [] })
  graph.edges.forEach(e => {
    if (adj[e.from]) adj[e.from].push({ to: e.to, weight: e.weight })
    if (adj[e.to]) adj[e.to].push({ to: e.from, weight: e.weight })
  })

  while (queue.length > 0) {
    const curr = queue.shift()

    steps.push({
      nodes: graph.nodes,
      edges: graph.edges,
      currentNode: curr,
      visitedNodes: Array.from(visited),
      visitedEdges: [...visitedEdges],
      queue: [...queue],
      line: 3,
      description: `Dequeued node ${curr}. Exploring neighbors...`
    })

    const neighbors = adj[curr] || []
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor.to)) {
        visited.add(neighbor.to)
        visitedEdges.push({ from: curr, to: neighbor.to })
        queue.push(neighbor.to)

        steps.push({
          nodes: graph.nodes,
          edges: graph.edges,
          currentNode: curr,
          visitedNodes: Array.from(visited),
          visitedEdges: [...visitedEdges],
          queue: [...queue],
          line: 5,
          description: `Discovered unvisited node ${neighbor.to} via edge (${curr}-${neighbor.to}). Enqueued ${neighbor.to}.`
        })
      }
    }
  }

  steps.push({
    nodes: graph.nodes,
    edges: graph.edges,
    currentNode: null,
    visitedNodes: Array.from(visited),
    visitedEdges: [...visitedEdges],
    queue: [],
    line: 8,
    description: `BFS Traversal Complete! All reachable nodes visited.`
  })

  return steps
}

export function generateDFSGraphSteps(graph = DEFAULT_GRAPH, startNode = 'A') {
  const steps = []
  const visited = new Set()
  const visitedEdges = []

  const adj = {}
  graph.nodes.forEach(n => { adj[n.id] = [] })
  graph.edges.forEach(e => {
    if (adj[e.from]) adj[e.from].push({ to: e.to, weight: e.weight })
    if (adj[e.to]) adj[e.to].push({ to: e.from, weight: e.weight })
  })

  steps.push({
    nodes: graph.nodes,
    edges: graph.edges,
    currentNode: startNode,
    visitedNodes: [],
    visitedEdges: [],
    line: 1,
    description: `Start DFS graph traversal from node ${startNode}`
  })

  function dfs(node) {
    visited.add(node)
    steps.push({
      nodes: graph.nodes,
      edges: graph.edges,
      currentNode: node,
      visitedNodes: Array.from(visited),
      visitedEdges: [...visitedEdges],
      line: 2,
      description: `Visiting node ${node}`
    })

    const neighbors = adj[node] || []
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor.to)) {
        visitedEdges.push({ from: node, to: neighbor.to })
        steps.push({
          nodes: graph.nodes,
          edges: graph.edges,
          currentNode: neighbor.to,
          visitedNodes: Array.from(visited),
          visitedEdges: [...visitedEdges],
          line: 4,
          description: `Traversing edge (${node}-${neighbor.to}) to visit node ${neighbor.to}`
        })
        dfs(neighbor.to)
      }
    }
  }

  dfs(startNode)

  steps.push({
    nodes: graph.nodes,
    edges: graph.edges,
    currentNode: null,
    visitedNodes: Array.from(visited),
    visitedEdges: [...visitedEdges],
    line: 7,
    description: `DFS Traversal Complete!`
  })

  return steps
}

export function generateDijkstraSteps(graph = DEFAULT_GRAPH, startNode = 'A') {
  const steps = []
  const dist = {}
  const visited = new Set()
  const visitedEdges = []

  graph.nodes.forEach(n => { dist[n.id] = Infinity })
  dist[startNode] = 0

  const adj = {}
  graph.nodes.forEach(n => { adj[n.id] = [] })
  graph.edges.forEach(e => {
    if (adj[e.from]) adj[e.from].push({ to: e.to, weight: e.weight })
    if (adj[e.to]) adj[e.to].push({ to: e.from, weight: e.weight })
  })

  steps.push({
    nodes: graph.nodes,
    edges: graph.edges,
    currentNode: startNode,
    visitedNodes: [],
    visitedEdges: [],
    distances: { ...dist },
    line: 1,
    description: `Initialize Dijkstra distances. ${startNode}=0, others=∞`
  })

  for (let i = 0; i < graph.nodes.length; i++) {
    let minDist = Infinity
    let u = null

    graph.nodes.forEach(n => {
      if (!visited.has(n.id) && dist[n.id] < minDist) {
        minDist = dist[n.id]
        u = n.id
      }
    })

    if (!u) break
    visited.add(u)

    steps.push({
      nodes: graph.nodes,
      edges: graph.edges,
      currentNode: u,
      visitedNodes: Array.from(visited),
      visitedEdges: [...visitedEdges],
      distances: { ...dist },
      line: 3,
      description: `Selected node ${u} with minimum distance ${minDist}`
    })

    const neighbors = adj[u] || []
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor.to)) {
        const alt = dist[u] + neighbor.weight
        if (alt < dist[neighbor.to]) {
          dist[neighbor.to] = alt
          visitedEdges.push({ from: u, to: neighbor.to })
          steps.push({
            nodes: graph.nodes,
            edges: graph.edges,
            currentNode: neighbor.to,
            visitedNodes: Array.from(visited),
            visitedEdges: [...visitedEdges],
            distances: { ...dist },
            line: 5,
            description: `Relaxed edge (${u}-${neighbor.to}): updated dist[${neighbor.to}] = ${alt}`
          })
        }
      }
    }
  }

  steps.push({
    nodes: graph.nodes,
    edges: graph.edges,
    currentNode: null,
    visitedNodes: Array.from(visited),
    visitedEdges: [...visitedEdges],
    distances: { ...dist },
    line: 8,
    description: `Dijkstra Shortest Path Computation Complete!`
  })

  return steps
}

export function generateAStarSteps(graph = DEFAULT_GRAPH) { return generateDijkstraSteps(graph) }
export function generatePrimsSteps(graph = DEFAULT_GRAPH) { return generateDijkstraSteps(graph) }
export function generateKruskalsSteps(graph = DEFAULT_GRAPH) { return generateDijkstraSteps(graph) }
